File: fui_system/src/platform/qt/qt_wrapper/qpixmap.rs
use crate::FUISystemError;

pub struct QPixmap {
    pub this: *mut ::std::os::raw::c_void,
}

impl QPixmap {
    pub fn from_data(data: &[u8]) -> Result<Self, FUISystemError> {
        unsafe {
            let this = crate::platform::qt::qt_wrapper::QPixmap_new();
            if this.is_null() {
                return Err(FUISystemError::OutOfMemory);
            }

            if crate::platform::qt::qt_wrapper::QPixmap_loadFromData(
                this,
                data.as_ptr(),
                data.len() as i32,
            ) == 0
            {
                return Err(FUISystemError::OsError(
                    "Cannot load pixmap from data.".to_string(),
                ));
            }

            Ok(Self { this })
        }
    }
}

impl Drop for QPixmap {
    fn drop(&mut self) {
        unsafe {
            crate::platform::qt::qt_wrapper::QPixmap_delete(self.this);
        }
    }
}

File: fui_system/src/platform/qt/qt_wrapper/cpp/qpixmap.cpp
#include <QPixmap>
#include "qpixmap.h"

void *QPixmap_new()
{
    return static_cast<void *>(new (std::nothrow) QPixmap());
}

void QPixmap_delete(void *self)
{
    delete static_cast<QPixmap *>(self);
}

int QPixmap_loadFromData(void *self, const unsigned char *data, int len)
{
    QPixmap *qPixmap = static_cast<QPixmap *>(self);
    return qPixmap->loadFromData((const uchar*)data, len);
}
